File: 210-course-schedule-ii/course-schedule-ii.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {number[]}
 */
var findOrder = function (numCourses, prerequisites) {
    const adj = new Array(numCourses).fill(null).map(() => new Array(0));
    for (let [a, b] of prerequisites) {
        adj[b].push(a);
    }

    const topoSort = getTopoSort(adj, numCourses);
    return topoSort.length !== numCourses ? [] : topoSort
};

var getTopoSort = function (adj, numCourses) {
    const indegrees = new Array(numCourses).fill(0);
    for (let edges of adj) {
        for (let node of edges) {
            indegrees[node]++;
        }
    }

    const queue = [];
    const stack = [];

    for (let i = 0; i < numCourses; i++) {
        if (indegrees[i] == 0) {
            queue.push(i);
        }
    }


    while (queue.length) {
        const node = queue.shift();
        stack.push(node);

        for (let nbr of adj[node]) {
            indegrees[nbr]--;
            if (indegrees[nbr] == 0) {
                queue.push(nbr);
            }
        }
    }

    return stack;
}

File: 210-course-schedule-ii/course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        // For (u,v) -> edge is v-->u

        vector<vector<int>> adj(numCourses);
        vector<int> indegree(numCourses, 0);

        for(auto& edge: prerequisites) {
            int v = edge[0], u = edge[1];
            adj[u].push_back(v);
            indegree[v]++; 
        }

        queue<int> q;
        for(int node = 0; node < numCourses; node++) {
            if(indegree[node] == 0) {
                q.push(node);
            }
        }

        vector<int> topo;

        while(!q.empty()) {
            int top = q.front();
            q.pop();
            topo.push_back(top);
            for(auto& nbr: adj[top]) {
                indegree[nbr]--;
                if(indegree[nbr] == 0) {
                    q.push(nbr);
                }
            }
        }

        return topo.size() == numCourses ? topo : vector<int>{};
    }
};